addons/payment_paypal/models/paypal_s2s.py

```diff
@@ -0,0 +1,231 @@
+            return True
+        elif response in ['pending', 'expired', 'failed', 'canceled']:
+            response = 'cancel' if response == 'canceled' else response
+            state_msg = 'Received response for Paypal payment '\
+                        'transaction %s: set as %s' % (self.reference, response)
+            _logger.info(state_msg)
+            self.write({
+                'state': response,
+                'acquirer_reference': data['id'],
+                'state_message': _(state_msg),
+            })
+            return True
+        else:
+            state_msg = 'Received unrecognized response for Paypal payment '\
+                        'transaction %s: set as %s' % (self.reference, response)
+            _logger.info(state_msg)
+            self.write({
+                'state': 'error',
+                'state_message': _(state_msg),
+            })
+            return True
+
+    @api.cr_uid_records_context
+    def _paypal_s2s_get_tx_status(self, cr, uid, tx, context=None):
+        headers = {
+            'Content-Type': 'application/json',
+            'Authorization': 'Bearer %s' % tx.acquirer_id._paypal_s2s_get_access_token()[tx.acquirer_id.id],
+        }
+        url = 'https://api.sandbox.paypal.com/v1/payments/payment/%s' % (tx.paypal_txn_id)
+        request = urllib2.Request(url, headers=headers)
+        data = self._paypal_try_url(request, tries=3, context=context)
+        return self.s2s_feedback(cr, uid, tx.id, data, context=context)
+
+
+class PaymentMethodPaypal(models.Model):
+    _inherit = 'payment.method'
+
+    @api.model
+    def paypal_create(self, values):
+
+        if values.get('cc_number'):
+            acquirer = self.env['payment.acquirer'].browse(int(values.get('acquirer_id')))
+
+            access_token = acquirer._paypal_s2s_get_access_token()
+            path = '/v1/vault/credit-card'
+            vault_url = self.env['payment.acquirer']._get_paypal_urls(acquirer.environment)['paypal_rest_url'] + path
+
+            vault_headers = {
+                'Authorization': 'Bearer %s' % access_token,
+                'Content-Type': 'application/json'
+            }
+            vault_params = {
+                'payer_id': values['partner_id'],
+                'type': values['cc_brand'],
+                'number': values['cc_number'].replace(' ', ''),
+                'expire_month': str(values['cc_expiry'][:2]),
+                'expire_year': '20' + str(values['cc_expiry'][-2:]),
+                'cvv2': values['cc_cvv'][:3],
+            }
+            vault_request = urllib2.Request(vault_url, json.dumps(vault_params), vault_headers)
+            vault_response = urllib2.urlopen(vault_request)
+            vault_result = json.loads(vault_response.read())
+            vault_response.close()
+
+            if not vault_result['state'] == 'ok':
+                _logger.error('Store credit card details on PayPal Vault has been failed')
```

Comment:
Use this please:
"Paypal Vault: error when storing credit card %s: %s" % (values['cc_number'][-4:], vault_result['state'])